sled-agent/src/rack_setup/service.rs

```diff
@@ -1084,6 +1066,123 @@ impl ServiceInner {
+    for sled_request in sled_plan.sleds.values() {
+        let sled_addr = get_sled_address(sled_request.body.subnet);
+        let sled_id = sled_request.body.id;
+        let entry = match sled_configs.entry(sled_id) {
+            btree_map::Entry::Vacant(entry) => entry,
+            btree_map::Entry::Occupied(_) => {
+                bail!(
+                    "duplicate sled address found while deriving blueprint: \
+                     {sled_addr}"
+                );
+            }
+        };
+        let sled_config =
+            service_plan.services.get(&sled_addr).with_context(|| {
+                format!(
+                    "missing services in plan for sled {sled_id} ({sled_addr})"
+                )
+            })?;
+        entry.insert(sled_config.clone());
+    }
+
+    Ok(build_initial_blueprint_from_sled_configs(
+        sled_configs,
+        internal_dns_version,
+    ))
+}
+
+pub(crate) fn build_initial_blueprint_from_sled_configs(
+    sled_configs: BTreeMap<Uuid, SledConfig>,
+    internal_dns_version: Generation,
+) -> Blueprint {
+    let mut omicron_zones = BTreeMap::new();
+    let mut zones_in_service = BTreeSet::new();
+    for (sled_id, sled_config) in sled_configs {
+        for zone in &sled_config.zones {
+            zones_in_service.insert(zone.id);
+        }
+        let zones_config = sled_agent_client::types::OmicronZonesConfig::from(
+            OmicronZonesConfig {
+                // This is a bit of a hack. We only construct a blueprint after
+                // completing RSS, so we need to know the final generation value
+                // sent to all sleds. Arguably, we should record this in the
+                // serialized RSS plan; however, we have already deployed
+                // systems that did not. We know that every such system used
+                // `V5_EVERYTHING` as the final generation count, so we can just
+                // use that value here. If we ever change this, in particular in
+                // a way where newly-deployed systems will have a different
+                // value, we will need to revisit storing this in the serialized
+                // RSS plan.
+                generation: DeployStepVersion::V5_EVERYTHING,
+                zones: sled_config.zones,
+            },
+        );
+        omicron_zones.insert(sled_id, zones_config);
+    }
+
+    Blueprint {
+        id: Uuid::new_v4(),
+        omicron_zones,
+        zones_in_service,
+        parent_blueprint_id: None,
+        internal_dns_version,
+        // We don't actually configure external DNS during RSS. During handoff,
+        // Nexus creates an initial empty external DNS configuration, then
+        // creates the recovery silo, which bumps the generation.
+        //
+        // TODO-john How should we do this? Baking this knowledge of how Nexus
+        // sets up external DNS here seems bad.
+        external_dns_version: Generation::new().next(),
```

Comment:
@davepacheco Do you have thoughts on this change I made getting caught up with `main`? A few obvious options:

* This is a little gross but not the end of the world
* We export an "external DNS generation after rack setup" from `nexus-db-queries` and use that here
* We set this to `Generation::new()` and let the planner sort out that there's a new external DNS later

Reply:
What about:
- we set it to `Generation::new()` here
- during rack initialization, Nexus sets it to whatever it wants after receiving this blueprint but before writing it to the database

The explanation would be: as far as RSS is concerned, it's generation 1 -- the oldest possible generation should be a safe bet and it doesn't know any better.  When Nexus receives it, it's in a position to say: "okay, RSS handed me this initial blueprint, but it didn't know I was going to write an initial external DNS generation, so I'm going to adjust that here based on the actions I'm taking [which are: creating generation 2 of external DNS]".  What do you think?

Reply:
I like that a lot - thanks! Done in [0b28298](https://github.com/oxidecomputer/omicron/pull/5244/commits/0b2829874a83445514e7124545f0866d796c8f3a)